content/post/difftree.md

```diff
@@ -0,0 +1,501 @@
+
+!["An example of a tree with some names and hashes in their
+nodes."](../../static/post/difftree/hashes.png "A typical Git tree, with some files and directories. The names of the nodes are shown between double quotes, the first bytes of their hashes are shown in blue.")
+
+This means Git trees are [Merkle
+trees](https://en.wikipedia.org/wiki/Merkle_tree), the practical implications of
+us are:
+
+- *Files with different contents have different hashes*.  This means it is
+  pretty fast to know if a file has changed between two commits, as you only
+  need to compare their hashes, not their contents.
+
+- *Two files with the same contents will have the same hash*, no matter if they
+  have different names and/or permissions.  If you index contents by hash, you
+  only need to store those contents once, no matter how many files hold those
+  contents in the history of the repository.
+
+- *If two directories have the same hash, then they have the same children*:
+  same files and directories, with the same names, modes and
+  contents, recursively.
+
+  This is a huge time saver when comparing directories between commits, because
+  if their hashes match, then you know they have not been modified; there is no
+  need to check their children individually.
+
+  Thanks to this, the worst case algorithmic complexity of checking two
+  directories for equality is reduced from O(n) to O(1) (with n being the
+  number of descendants of the directory with less descendants).
+
+- If some change has been introduced in a commit, no matter how small, all the
+  ancestor directories of the modified file will see their hashes modified.
+
+Let us assume, for the sake of this blog post, that hash collisions are
+impossible, even though they are quite popular
+[nowadays](https://shattered.io/static/shattered.pdf)).
+
+# Representing Git trees programmatically
+
+In line with what we have exaplined so far, this is a sensible representation of
+Git trees in Go:
+
+```go
+type Noder interface {
+        Name()     string
+        Hash()     []byte
+        Children() []Noder
+}
+```
+
+The path to a node in a tree will look something like this:
+
+```go
+type Path []Noder // beginning at the root and ending with the node itself
+```
+
+I recommend `Path` implements `Noder` so you can treat paths as
+nodes when needed:
+
+```go
+// assuming the receiver is not the empty slice…
+func (p Path) last() Noder       { return p[len(p)-1] }
+func (p Path) Name() string      { return p.last().Name() }
+func (p Path) Hash() []byte      { return p.last().Hash() }
+func (p Path) Children() []Noder { return p.last().Children() }
+```
+
+# Representing changes in Git trees
+
+Let A and B be two Git trees from consecurive commits.  If we compare one with
+the other there is going to be quite a few kind of changes we will have to deal
+with, let us see a minimal set (changes depicted in red):
+
+- **Modified file**: Both trees will have the same topology, but the hash of the
```

Comment:
~will~
The reason we shouldn't use the future tense here is because we describe the usual way things happen. If a file changes, the trees have the same topology but blah-blah-blah. ""If you heat up the water, it boils."